File: GroceryStore.Web/src/app/services/product.service.ts
import { Injectable, signal } from '@angular/core';
import { IProductService } from '../interfaces/product-service.interface';
import { forkJoin, map, Observable } from 'rxjs';
import { CreateProductRequest, GetInventoryOverviewRequest, UpdateProductRequest } from '../models/requests/product-requests';
import { CreateProductResponse, GetProductResponse, GetProductsResponse, UpdateProductResponse, DeleteProductResponse, GetInventoryOverviewResponse } from '../models/responses/product-responses';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../environments/environment';
import { Product } from '../models/product';
import { AisleService } from './aisle.service';

@Injectable({
  providedIn: 'root'
})

export class ProductService implements IProductService{
    private apiUrl = `${environment.apiUrl}/products`

    private productsSignal = signal<Product[] | undefined>(undefined)

    private today: Date = new Date();
    private oneWeekAgo: Date = new Date();
    private oneWeekAhead: Date = new Date();

    constructor(private http: HttpClient, private aisleService: AisleService) { 
        this.oneWeekAgo.setDate(this.today.getDate() - 7)
        this.oneWeekAhead.setDate(this.today.getDate() + 7)
    }

    createProduct(request: CreateProductRequest): Observable<CreateProductResponse> {
        return this.http.post<CreateProductResponse>(`${this.apiUrl}/create`, request);
    }

    getProduct(id: string): Observable<GetProductResponse> {
        return this.http.get<GetProductResponse>(`${this.apiUrl}/${id}`);
    }

    getProducts(): Observable<GetProductsResponse> {
        // Note: Caching and joins in the back end were considered, but not implemented due to project scope.
        // For demo purposes, joining is handled on the front end, which is acknowledges as less efficient.
        // Further optimization could invvolve implementing joining in the back end/db, and caching/cache
        // busting in the front end.
        return forkJoin({
            aisles: this.aisleService.getAisles(),
            products: this.http.get<GetProductsResponse>(`${this.apiUrl}`)
        }).pipe(
            map(({aisles: aisleResponse, products: productResponse}) => {
                const aisleMap = new Map<string, string>();

                aisleResponse.aisles.forEach(aisle => {
                    aisleMap.set(aisle.id, aisle.name);
                });

                const products = productResponse.products.map(product => {
                    product.aisleName = aisleMap.get(product.aisleId) || 'Unknown';
                    product.createdOn = new Date(product.createdOn);
                    product.purchaseDate = new Date(product.purchaseDate);
                    product.expiryDate = new Date(product.expiryDate);
                    if (product.modifiedOn) {
                        product.modifiedOn = new Date(product.modifiedOn);
                    }
                    return product;
                });

                this.productsSignal.set(products);
                return productResponse;
            })
        );
    }

    getProductsByAisleId(aisleId: string): Observable<Product[]> {
        return this.getProducts().pipe(
            map(response => response.products.filter(product => aisleId === product.aisleId))
        );
    }
    
    updateProduct(request: UpdateProductRequest): Observable<UpdateProductResponse> {
        return this.http.put<UpdateProductResponse>(`${this.apiUrl}/update`, request)
    }

    deleteProduct(id: string): Observable<DeleteProductResponse> {
        return this.http.delete<DeleteProductResponse>(`${this.apiUrl}/delete/${id}`)
    }

    getInventoryOverview(request: GetInventoryOverviewRequest): Observable<GetInventoryOverviewResponse> {
        const params = new HttpParams()
        .set('purchaseDate', request.purchaseDate.toISOString())
        .set('expiryDate', request.expiryDate.toISOString());

        return this.http.get<GetInventoryOverviewResponse>(`${this.apiUrl}/overview`, {params});
    }
}

File: GroceryStore.API/Services/ProductService.cs
using GroceryStore.API.Exceptions.Aisle;
using GroceryStore.API.Exceptions.Product;
using GroceryStore.API.Models;
using GroceryStore.API.Models.Requests.Product;
using GroceryStore.API.Models.Responses.Product;
using GroceryStore.API.Repositories.Interfaces;
using GroceryStore.API.Services.Interfaces;
using System.Linq;
using System.Threading.Tasks;

namespace GroceryStore.API.Services
{
    public class ProductService : IProductService
    {
        private readonly IProductRepository _productRepository;
        private readonly IAisleRepository _aisleRepository;

        public ProductService(IProductRepository productRepository, IAisleRepository aisleRepository)
        {
            _productRepository = productRepository;
            _aisleRepository = aisleRepository;
        }

        public async Task<CreateProductResponse> CreateProduct(CreateProductRequest request)
        {
            var resp = await _aisleRepository.GetAisleById(request.AisleId);

            if (resp.Aisle == null)
                throw new AisleNotFoundException(request.AisleId);

            return await _productRepository.CreateProduct(request);
        }

        public async Task<GetProductResponse> GetProductById(string id)
        {
            var resp = await _productRepository.GetProductById(id);

            if (resp.Product == null)
                throw new ProductNotFoundException(id);

            return resp;
        }

        public async Task<GetProductsResponse> GetProductByAisleId(string aisleId)
        {
            var aisleResponse = await _aisleRepository.GetAisleById(aisleId);

            if (aisleResponse.Aisle == null)
                throw new AisleNotFoundException(aisleId);

            var resp = await _productRepository.GetProductsByAisleId(aisleId);

            if (resp.Products == null)
                throw new ProductNotFoundException();

            return resp;
        }

        public async Task<GetProductsResponse> GetProducts()
        {
            return await _productRepository.GetProducts();
        }

        public async Task<GetProductsWithAislesResponse> GetProductsWithAisles()
        {
            var products = await _productRepository.GetProducts();

            var aisles = await _aisleRepository.GetAisles();


            // Note: More performant joins were considered but not implemented due to project scope.
            // For demo purposes, filtering is handled on the front end, which is acknowledged as less efficient. 
            // Further optimization could involve implementing backend pagination and filtering if required.
            var response = new GetProductsWithAislesResponse(products.Products, aisles.Aisles);

            return response;
        }

        public async Task<UpdateProductResponse> UpdateProduct(UpdateProductRequest request)
        {
            var resp = await _productRepository.GetProductById(request.Id);

            if (resp.Product == null)
                throw new ProductNotFoundException(request.Id);

            return await _productRepository.UpdateProduct(request);
        }

        public async Task<DeleteProductResponse> DeleteProduct(string id)
        {
            var resp = await _productRepository.GetProductById(id);

            if (resp.Product == null)
                throw new ProductNotFoundException(id);

            return await _productRepository.DeleteProduct(id);
        }

        public async Task<GetInventoryOverviewResponse> GetInventoryOverview(GetInventoryOverviewRequest request)
        {
            var productQuery = await _productRepository.GetProducts();

            var aisleQuery = await _aisleRepository.GetAisles();

            var response = new GetInventoryOverviewResponse {
                TotalProducts = productQuery.Total,
                TotalAisles = aisleQuery.Aisles.Count,
                ProductOverviews = productQuery.Products
                                    .GroupBy(p => p.Name)
                                    .Select(group => new ProductOverview() {
                                        Name = group.Key,
                                        SumProducts = group.Sum(p => p.Quantity),
                                        SumExpiring = group.Where(p => p.ExpiryDate <= request.ExpiryDate).Sum(p => p.Quantity),
                                        SumPurchased = group.Where(p => p.PurchaseDate >= request.PurchaseDate).Sum(p => p.Quantity)
                                    }).ToList()
            };

            return response;
        }
    }
}
